src/getindex.jl

```diff
@@ -0,0 +1,143 @@
+Base.checkbounds(A::LinearMap, I::AbstractMatrix{Bool}) =
+    axes(A) == axes(I) || throw(BoundsError(A, I))
+
+# main entry point
+function Base.getindex(A::LinearMap, I...)
+    # TODO: introduce some sort of switch?
+    @boundscheck checkbounds(A, I...)
+    _getindex(A, Base.to_indices(A, I)...)
+end
+# quick pass forward
+Base.@propagate_inbounds Base.getindex(A::ScaledMap, I...) = A.λ .* A.lmap[I...]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, I...) = A.lmap[I...]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, i::Integer) = A.lmap[i]
+Base.@propagate_inbounds Base.getindex(A::WrappedMap, i::Integer, j::Integer) = A.lmap[i,j]
+
+########################
+# linear indexing
+########################
+function _getindex(A::LinearMap, i::Integer)
+    i1, i2 = Base._ind2sub(axes(A), i)
+    return _getindex(A, i1, i2)
+end
+_getindex(A::LinearMap, I::Indexer) = [_getindex(A, i) for i in I]
+_getindex(A::LinearMap, ::Base.Slice) = vec(Matrix(A))
+_getindex(A::LinearMap, I::Vector{CartesianIndex{2}}) = [(@inbounds A[i]) for i in I]
+_getindex(A::LinearMap, I::Base.LogicalIndex) = [(@inbounds A[i]) for i in I]
+
+########################
+# Cartesian indexing
+########################
+_getindex(A::LinearMap, i::Union{Integer,Indexer}, j::Integer) = (@inbounds (A*basevec(A, 2, j))[i])
+_getindex(A::LinearMap, ::Base.Slice, j::Integer) = A*basevec(A, 2, j)
+function _getindex(A::LinearMap, i::Integer, J::Indexer)
+    # try
+        # requires adjoint action to be defined
+        return @inbounds (basevec(A, 1, i)'A)[J]
+    # catch
+    #     return _fillbycols!(zeros(eltype(A), Base.index_shape(i, J)), A, i, J)
+    # end
+end
+function _getindex(A::LinearMap, i::Integer, J::Base.Slice)
+    # try
+        # requires adjoint action to be defined
+        return vec(basevec(A, 1, i)'A)
+    # catch
+    #     return _fillbycols!(zeros(eltype(A), Base.index_shape(i, J)), A, i, J)
+    # end
+end
+function _getindex(A::LinearMap, I::Indexer, J::Indexer)
+    dest = zeros(eltype(A), Base.index_shape(I, J))
+    if length(I) <= length(J)
+        # try
+            # requires adjoint action to be defined
+            _fillbyrows!(dest, A, I, J)
+        # catch
+        #     _fillbycols!(dest, A, I, J)
+        # end
+    else
+        _fillbycols!(dest, A, I, J)
+    end
+    return dest
+end
+_getindex(A::LinearMap, ::Base.Slice, ::Base.Slice) = Matrix(A)
+
+# specialized methods
+_getindex(A::FillMap, ::Integer, ::Integer) = A.λ
+_getindex(A::LinearCombination, i::Integer, j::Integer) =
+    sum(a -> (@inbounds A.maps[a][i, j]), eachindex(A.maps))
+_getindex(A::AdjointMap, i::Integer, j::Integer) = @inbounds adjoint(A.lmap[j, i])
+_getindex(A::TransposeMap, i::Integer, j::Integer) = @inbounds transpose(A.lmap[j, i])
+_getindex(A::UniformScalingMap, i::Integer, j::Integer) = ifelse(i == j, A.λ, zero(eltype(A)))
+
+# helpers
+function basevec(A, dim, i::Integer)
```

Comment:
Mathematically this is a "standard unit vector" right?   So `unitvec` would be a logical name?
Never mind, this is bike shedding. :)

Reply:
True, I'll go with that.